### PR TITLE
DPL: keep inputTimeslice into account when enumerating

### PR DESCRIPTION
**Pull request by ktf**

### CODE REVIEW

**Comment by ktf** [comment]
@shahor02 can you check if this (in addition to the other PR) allows you to remove the extra device?

File: Framework/Core/src/LifetimeHelpers.cxx [hunk 1]
```diff
@@ -275,6 +276,11 @@ ExpirationHandler::Handler LifetimeHelpers::enumerate(ConcreteDataMatcher const&
 {
   auto counter = std::make_shared<int64_t>(0);
   auto f = [matcher, counter, sourceChannel](ServiceRegistry& services, PartRef& ref, uint64_t timestamp) -> void {
+    // We generate entries according to our inputTimeslice.
+    if ((*counter % services.get<DeviceSpec>().maxInputTimeslices) != services.get<DeviceSpec>().inputTimesliceId) {
```
Comment (shahor02) [inline Framework/Core/src/LifetimeHelpers.cxx]: There should be ``services.get<const DeviceSpec>`` here, I am getting:
``[ERROR] Unhandled exception reached the top of main: Unable to find service of kind N2o29framework10DeviceSpecE did you register one as non-const reference?, device shutting down.``

**Comment by shahor02** [comment]
@ktf  with the fix I above this PR makes also the pipeline w/o upstream device to stagger the TFs and provide correct ``inputTimesliceId``. 
But the processing gets stuck after a few 10 timeframes, w/o any message. The exact number of TFs it produces is reproducible and somehow depends on the number of lanes in the pipeline, but I did not catch any regularity, should be something modulo something...

Then I picked this PR to the workflow with the device upstream, now I get consistently 2000 frames (i.e. the highest ``DataProcessingHeader. startTime`` is 2000*nlanes) processed per lane, then it also gets stuck. I've gradually reverted 1st this then the previous PR https://github.com/AliceO2Group/AliceO2/pull/2981, in both cases I still observe the same behaviour. 
Finally, I've removed the ``timePipeline`` adding the devine to the workflow, from https://github.com/shahor02/AliceO2/blob/96360bb0e55b92e21ec4b2e8f621207f784c7e55/Detectors/TOF/calibration/testWorkflow/data-generator-workflow.cxx#L32
and I still get only 2000 frames processed. 
I did not find any hardcoded limit in the Framework and I am sure don't have it either, so I am lost.
Any hint?

**Comment by ktf** [comment]
You are of course right for the `const`.

There shouldn't be any limit, let's check together when we get to CERN.

**Comment by ktf** [comment]
@shahor02 this should remove the need for the intermediate device.
@knopers8 I assume you are not affected by this because you always had a non-timepipelined device before a set of time-pipelined ones. Can you otherwise check you have the expected behavior?
@knopers8 what would you expect for a timer? Same behaviour as now or the one of the pull request?

**Comment by shahor02** [comment]
@ktf thanks, I confirm that now everything works as expected, both with and w/o an upstream device.

**Comment by ktf** [comment]
Current behaviour is in any case wrong. Merging this.

**Comment by knopers8** [comment]
That shouldn't what I have so far. I don't understand though why it is:
```
auto last = std::make_shared<size_t>(start);
...
*last += step * (maxInputTimeslices + inputTimeslice);
```
Instead of:
```
auto last = std::make_shared<size_t>(start + inputTimeslice * step);
...
*last += step * maxInputTimeslices;
```
As far as I understand, if we have three devices and step 1, device_t0 should produce 0, 3, 6..., device_t1: 1, 4, 7... and device_t2: 2, 5, 8...
With this enumeration don't we get this?
device_t0: 0, 3, 6, 9...
device_t1: 0, 4, 8, 12...
device_t2: 0, 5, 10, 15...
Maybe I am missing something here.

For timers, I guess it would be good to have a mechanism which avoids collisions of timesliceIDs generated by timers. I am not sure though, if stopping to use current time and enumerating timesliceIDs there would break something.

**Comment by ktf** [comment]
yes, you are of course right.

**Comment by ktf** [comment]
Should be fixed in #3001

**Pull request merged**